ibis/common/graph.py

```diff
@@ -0,0 +1,151 @@
+"""
+Various traversal utilities for the expression graph.
+"""
+from __future__ import annotations
+
+from abc import abstractmethod
+from collections import deque
+from typing import Any, Callable, Iterable, Iterator
+
+import toolz
+
+# TODO(kszucs): add recursive implementation for bfs and dfs
+
+
+class Traversable:
+
+    __slots__ = ()
+
+    @property
+    @abstractmethod
+    def __children__(self):
+        ...
+
+
+def bfs(node: Traversable, types=Traversable):
+    assert issubclass(types, Traversable)
+    assert isinstance(node, types)
+
+    queue = deque([node])
+    graph = dict()
+
+    while queue:
+        node = queue.popleft()
+        children = [c for c in node.__children__ if isinstance(c, types)]
+        graph[node] = children
+        queue.extend(children)
+
+    return graph
+
+
+def dfs(node: Traversable, types=Traversable):
+    assert issubclass(types, Traversable)
+    assert isinstance(node, types)
+
+    stack = deque([node])
```

Comment:
No reason to use a `deque` here.

```suggestion
    stack = [node]
```

Reply:
deque is supposed to be slightly more performant in this scenario due to frequent subsequend adding and popping